File: CrackingTheCodingInterview/Chapter2/two-eight.py
# find a loop in a linked list

# passes in the root node
def loop(root):
    fast = root
    slow = root
    
    while(fast.next != None and fast != None):
        fast = fast.next.next
        slow = slow.next
        if (slow == fast):
            break
    
    if(fast == None or fast.next == None):
        return None

    fast = root
    while(slow != fast):
        slow = slow.next
        fast = fast.next
    return slow

        

File: CrackingTheCodingInterview/Chapter2/TwoEight.java
package CrackingTheCodingInterview.Chapter2;

// finds loop in linked list
public class TwoEight 
{
    LinkedListNode loop(LinkedListNode root)
    {
        LinkedListNode fast = root;
        LinkedListNode slow = root;
        while (fast != null && fast.getNext() != null)
        {
            fast = fast.getNext().getNext();
            slow = slow.getNext();
            if (fast == slow)
            {
                break;
            }
        }

        if (fast == null || fast.getNext() == null)
        {
            return null;
        }

        fast = root;
        while (fast != slow)
        {
            fast = fast.getNext();
            slow = slow.getNext();
        }

        return fast;
    }

}